### ISSUE TITLE
Per channel loss?

### ISSUE THREAD

**Issue by FrostedPenguin** [fchollet/keras#8403]
Hello. Currently I'm building autoencoder. (RGB->RGB)
However my data is not actually color image. Other type of data is encoded in each channel. 
Domain of each channels are different and the loss function of each channel should be different.
For example I can imagine the loss function looks like this.
```
def CustomLoss(y_true, y_pred):
    y_true_R = y_true[:,:,:,0]
    y_true_G = y_true[:,:,:,1] * 2.0 - 1.0 
    y_true_B = log(y_true[:,:,:,2])
    
    y_pred_R = y_pred[:,:,:,0]
    y_pred_G = y_pred[:,:,:,1] * 2.0 - 1.0
    y_pred_B = log(y_pred[:,:,:,2])

    Loss_R = l2(y_trueR, y_predR)
    Loss_G = mean_absolute_percentage(y_trueR, y_predR)
    Loss_B = l1(y_trueR, y_predR)
    return Loss_R * 0.3 + Loss_G * 0.5 + Loss_B * 0.2
```
I can simply build multi output model. However format of my prediction should be RGB and I need to prep per channel images for each losses. That is cumbersome. Also the above approach feels more simple.

So my question is that is it possible to use that kind of loss function?
And would it shows difference compare to multi output model?

**Comment by thefiddler** [comment]
> So my question is that is it possible to use that kind of loss function?
Yes, this is a valid approach for a loss function.

> And would it shows difference compare to multi output model?
No, both approaches (multi-output vs RGB output) will give the same results, as long as you don't have a bug in your code. Feel free to use whichever approach feels more intuitive for your use case.

> However format of my prediction should be RGB
In that case, the simplest approach is to use a single 3-channel output and train with your custom loss function.

You can still get the right results with a 3-output model, but you'd have to add one more layer to your model after training:
```python
# for training
training_model = create_model(...) # x -> [y1, y2, y3]
training_model.compile(loss=[y1_loss, y2_loss, y3_loss], ...)
training_model.train(...)

# for predictions
x = Input(shape=(1, None, None))
y1, y2, y3 = training_model(x)
y_rgb = concatenate([y1, y2, y3], axis=1)
model = Model([x], [y_rgb])
model.compile()
result = model.predict(...)
```

**Comment by FrostedPenguin** [comment]
Fully understood. Thank you for your detailed explanation.

**Comment by FrostedPenguin** [comment]
@thefiddler 
I just look into your code and found
```
y_rgb = concatenate([y1, y2, y3], axis=1) # combine 3 outputs to a single RGB output
```
To concat image channels, I always used axis=-1. 
But why did you use axis=1? Is there any reason for that?
